File: test/calculator/clnt.jl
using Thrift
import Thrift.process, Thrift.meta

include("gen-jl/floatops/floatops_constants.jl");
include("gen-jl/floatops/floatops_types.jl");
include("gen-jl/floatops/FloatCalc.jl");

include("gen-jl/arithmetic/arithmetic_constants.jl");
include("gen-jl/arithmetic/arithmetic_types.jl");
include("gen-jl/arithmetic/Calc.jl");

function calcclnt(niter::Int)
    clnt_transport = TSocket(19999)
    #proto = TCompactProtocol(clnt_transport)
    proto = TBinaryProtocol(clnt_transport)

    clnt = CalcClient(proto)

    println("opening connection. $(clnt_transport)")
    open(clnt_transport)
    println("opened connection. $(clnt_transport)")
    nops = length(OPS)
    for idx in 1:niter
        try
            op = string(OPS[round(Int, (nops-1)*rand())+1])
            p1 = round(Int32, 100*rand())
            p2 = round(Int32, 100*rand())
            res = calculate(clnt, op, p1, p2)
            println("$op ( $p1, $p2 ) = $res")
        catch ex
            !isa(ex, InvalidOperation) && rethrow()
            println(ex.oper)
        end
    end
    close(clnt_transport)
    println("closed connection. $(clnt_transport)")
end

function floatcalcclnt(niter::Int)
    clnt_transport = TSocket(19999)
    #proto = TCompactProtocol(clnt_transport)
    proto = TBinaryProtocol(clnt_transport)

    clnt = CalcClient(proto)

    println("opening connection. $(clnt_transport)")
    open(clnt_transport)
    println("opened connection. $(clnt_transport)")
    nops = length(FLOAT_OPS)
    for idx in 1:niter
        try
            m = (idx > 3) ? 50 : 8
            op = string(FLOAT_OPS[round(Int, (nops-1)*rand())+1])
            p1 = (2^m) * randn()
            p2 = (2^m) * randn()
            res = float_calculate(clnt, op, p1, p2)
            println("$op ( $p1, $p2 ) = $res")
        catch ex
            !isa(ex, InvalidFloatOperation) && rethrow()
            println(ex.oper)
        end
    end
    close(clnt_transport)
    println("closed connection. $(clnt_transport)")
end

for idx in 1:5
    calcclnt(10)
    floatcalcclnt(10)
end

File: test/calculator/clnt.py
#!/usr/bin/env python

import sys, random
sys.path.append('./gen-py')

from arithmetic import Calc
from arithmetic.ttypes import *
from arithmetic.constants import *

from floatops.ttypes import *
from floatops.constants import *

from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol, TCompactProtocol

def calcclnt(niter):
    try:
        # Make socket
        transport = TSocket.TSocket('localhost', 19999)

        # Buffering is critical. Raw sockets are very slow
        transport = TTransport.TBufferedTransport(transport)

        # Wrap in a protocol
        protocol = TBinaryProtocol.TBinaryProtocol(transport)
        #protocol = TCompactProtocol.TCompactProtocol(transport)

        # Create a client to use the protocol encoder
        client = Calc.Client(protocol)

        # Connect!
        print("opening connection. " + str(transport))
        transport.open()
        print("opened connection. " + str(transport))

        for idx in range(0, niter):
            try:
                oper = random.choice(OPS)
                p1 = random.randint(1, 100)
                p2 = random.randint(1, 100)
                ret = client.calculate(oper, p1, p2)
                print("%s(%d, %d) = %d"%(oper, p1, p2, ret))
            except InvalidOperation, tx:
              print "%s" % (tx.oper)

        transport.close()
        print("closed connection. " + str(transport))

    except Thrift.TException, tx:
      print "%s" % (tx.message)

def floatcalcclnt(niter):
    try:
        # Make socket
        transport = TSocket.TSocket('localhost', 19999)

        # Buffering is critical. Raw sockets are very slow
        transport = TTransport.TBufferedTransport(transport)

        # Wrap in a protocol
        protocol = TBinaryProtocol.TBinaryProtocol(transport)
        #protocol = TCompactProtocol.TCompactProtocol(transport)

        # Create a client to use the protocol encoder
        client = Calc.Client(protocol)

        # Connect!
        print("opening connection. " + str(transport))
        transport.open()
        print("opened connection. " + str(transport))

        for idx in range(0, niter):
            try:
                oper = random.choice(FLOAT_OPS)
                m = 50 if (idx > 3) else 8
                p1 = random.uniform(1, 2**m)
                p2 = random.uniform(1, 2**m)
                ret = client.float_calculate(oper, p1, p2)
                print("%s(%f, %f) = %f"%(oper, p1, p2, ret))
            except InvalidFloatOperation, tx:
              print "%s" % (tx.oper)

        transport.close()
        print("closed connection. " + str(transport))

    except Thrift.TException, tx:
      print "%s" % (tx.message)

for idx in range(0,5):
    calcclnt(10)
    floatcalcclnt(10)

